### DISCUSSION TITLE
How to do unit testing when using Axios instances?

### DISCUSSION THREAD

**Discussion by victorhcortesp** [mswjs/msw#1254] [Q&A]
I am having trouble getting `msw` to intercept my requests.

I have a configuration file in the root, to get all the environment variables. Something like this
- config.js

```javascript
// config.js

const { BASE_URL } = process.env;

const config = { baseUrl: BASE_URL };

module.exports = config;
```

I have a folder called mocks and inside it I have two files
- handlers.js
NOTE: In the handlers I am performing a stub, the value returned by the config as this baseURL can change depending on the environment variable, so the stub is to ensure that the instance always takes this default value `BASE_URL`.

```javascript
// handlers.js

const { stub } = require('sinon');
const { rest } = require('msw');
const config = require('../config');

const BASE_URL = 'http://localhost:8000';
stub(config, 'baseUrl').value(BASE_URL);

const handlers = [
  rest.post(`${BASE_URL}/positions`, (req, res, ctx) => {
    return res(
      ctx.status(200),
      ctx.json({
        id: 'f79e82e8-c34a-4dc7-a49e-9fadc0979fda',
      })
    );
  })
];

module.exports = { handlers };
```
- server.js
```javascript
// server.js

const { setupServer } = require('msw/node');
const { handlers } = require('./handlers');

const server = setupServer(...handlers);

module.exports = { server };
```

I have another file at a higher level where I create some instances and some functions that use those instances to make the requests. Something like this
```javascript
// instance.js

const axios = require('axios');
const { baseUrl } = require('./config');

const instanceAxios = axios.create({ baseURL: baseUrl });

const addPosition = async (data) => {
  const response = await instanceAxios.post('/positions', data);
  return response.data;
};

module.exports = { addPosition }
```

From this last file I want to do the unit tests, I am trying something like this, but I am not getting the intersection.
```javascript
// instance.test.js

const { server } = require('./mocks/server.js');
const { addPosition } = require('./instance.js');

describe('Axios API', () => {
  beforeAll(() =>
    server.listen({
      onUnhandledRequest: 'warn',  // I added this but it doesn't show me anything, as if it was detecting the route correctly.
    })
  );

  afterEach(() => server.resetHandlers());

  afterAll(() => server.close());

  it('should call axios and return a response', async () => {
    server.printHandlers();  // This is showing me the route configurations fine.
    
    const data = { username: 'username', firstName: 'firstName', lastName: 'lastName' };

    const response = await addPosition(data);

    console.log(response);  // I just want to validate if it is returning the mocked value.
  });
});

```
you can imagine a folder structure similar to this one

```
instance.js
instance.test.js
config.js
mocks
├── server.js
├── handlers.js
```

I would like to know what I am doing wrong or if it is definitely some `msw` problem that does not work very well with axios instances.

## Environment
msw: 0.42.0
nodejs: 16.15.0
npm: 8.5.5
axios: 0.27.2

**Comment by kettanaito** [comment]
Hey, @victorhcortesp. Thanks for starting this discussion.

At first glance, I see nothing wrong with your MSW setup. Could you please create a reproduction repository for me to look at? Then I'd be able to help you much more efficiently. Without it, it's just wild guessing. 

We do have a few issues reported around Axios and the latest version of MSW, so this may or may not be the case. In those issues, however, Axios usage threw errors, which doesn't seem to happen in your case. It'd be nice to confirm that.

**Reply by victorhcortesp** [reply]
Hi, thank you very much, I will repository a repo and share it.

**Reply by victorhcortesp** [reply]
Hi @kettanaito, this is the [repository](https://github.com/victorhcortesp/Example-Mock-Service-Worker)

**Comment by victorhcortesp** [comment]
Hi @kettanaito , I have made several tests but I can't find the error, I don't know if it is a failure of my implementation or if it is an issue of the library.

**Reply by kettanaito** [reply]
Hi, @victorhcortesp. Can you please revert to MSW 0.39.x and run your test suite again? Does that change anything?

**Reply by victorhcortesp** [reply]
Hi @kettanaito, Testing with those versions and nothing. I still have the error

**Reply by kettanaito** [reply]
Thanks for giving it an A/B check 👍

**Comment by victorhcortesp** [comment]
@kettanaito I was testing and I realized that the post is not working properly, I don't know where I have the error.


```javascript
const { rest } = require('msw');
const mswNode = require('msw/node');
const axios = require('axios');

const BASE_URL = 'http://localhost:8000';

const server = mswNode.setupServer(
  rest.get(`${BASE_URL}/users`, (_, res, ctx) => {
    return res(
      ctx.status(200),
      ctx.json({
        username: 'username',
        firstName: 'firstName',
        lastName: 'lastName',
      })
    );
  }),
  rest.post(`${BASE_URL}/user`, (_req, res, ctx) => {
    return res(
      ctx.status(200),
      ctx.json({
        id: 'f79e82e8-c34a-4dc7-a49e-9fadc0979fda',
      })
    );
  })
);

describe('Axios API', () => {
  beforeAll(() =>
    server.listen({
      onUnhandledRequest: 'warn', // I added this but it doesn't show me anything, as if it was detecting the route correctly.
    })
  );

  afterEach(() => server.resetHandlers());

  afterAll(() => server.close());

  it('should call axios get and return a response', async () => {
    const data = {
      username: 'username',
      firstName: 'firstName',
      lastName: 'lastName',
    };

    const response = await axios.get(`${BASE_URL}/users`);
    expect(response).toMatchObject({ data, status: 200 });
  });

  it('should call axios post and return a response', async () => {
    server.printHandlers(); // This is showing me the route configurations fine.

    const data = {
      username: 'username',
      firstName: 'firstName',
      lastName: 'lastName',
    };

    const response = await axios.post(`${BASE_URL}/user`, data);
    console.log(response.data);
  });
});

```

This is the result of the tests with the previous file
<img width="817" alt="image" src="https://user-images.githubusercontent.com/38927248/172493609-a4a8e389-6398-4dfe-886a-c2faf817ab4c.png">

**Comment by victorhcortesp** [comment]
@kettanaito I did some tests with PUT, DELETE passing `:id` in the URL and it works, the problem is with the post.
<img width="567" alt="image" src="https://user-images.githubusercontent.com/38927248/172497575-7acb7831-2a5e-4ea1-af99-ea4852bfacb3.png">

**Comment by victorhcortesp** [comment]
@kettanaito The failure occurs when I pass the second parameter, if I remove it, the test works but fails with the second parameter.

```javascript
const response = await axios.post(`${BASE_URL}/users`, data)
```

**Comment by kettanaito** [comment]
I think this is connected to https://github.com/mswjs/msw/issues/1125, which is fixed by https://github.com/mswjs/interceptors/pull/260. The release should happen tonight, so give it a try following the instructions in 1125.

**Reply by victorhcortesp** [reply]
Hi @kettanaito, With the new version 0.42.1 the tests I am performing are working fine.

Thank you very much for your help.

**Reply by ashwary23** [reply]
hey @kettanaito , the post-call is still not intercepting if I am passing the second param
ref:

`function submitContractName(data: Object) {
  return axiosInstance.post(
    `contract-management/v1/contracts/contract-name`,
    data
  );
}

handler.ts
http.post("contract-management/v1/contracts/contract-name", async ({ request }) => {
(never reaches here)
      console.log("create contract", request.json());
    }),
`
packages:
msw: 2.0.11
node: 18.17.1
"jest": "^29.7.0",
